File: app/src/main/java/com/xtr/keymapper/MainActivity.java
package com.xtr.keymapper;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.content.res.ColorStateList;
import android.os.Bundle;
import android.provider.Settings;
import android.widget.Button;


public class MainActivity extends AppCompatActivity {
    public static final int DEFAULT_PORT = 6234;
    public static final int DEFAULT_PORT_2 = 6345;
    public TouchPointer pointerOverlay;
    public Server server;

    private Button startOverlayButton;
    private Button startServerButton;
    private Button startInTerminal;
    private Button keymap;
    private Button configureButton;
    private Button infoButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        server = new Server(this);
        pointerOverlay = new TouchPointer(this);
        new Thread(pointerOverlay::startSocket).start();
        initButtons(); setupButtons();
    }

    private void setupButtons() {
        startServerButton.setOnClickListener(v -> startServer(true));
        startInTerminal.setOnClickListener(v -> startServer(false));
        startOverlayButton.setOnClickListener(v -> startService());
        keymap.setOnClickListener(v -> startEditor());
        configureButton.setOnClickListener(v -> startActivity(new Intent(this, InputDeviceSelector.class)));
        infoButton.setOnClickListener(v -> startActivity(new Intent(this, InfoActivity.class)));
    }

    private void initButtons(){
        startOverlayButton = findViewById(R.id.startPointer);
        startServerButton = findViewById(R.id.startServer);
        startInTerminal = findViewById(R.id.startServerM);
        keymap = findViewById(R.id.start_editor);
        configureButton = findViewById(R.id.config_pointer);
        infoButton = findViewById(R.id.about_button);
    }

    private void startService(){
        checkOverlayPermission();
        if(Settings.canDrawOverlays(this)) {
            pointerOverlay.open();
        }
    }

    public void setButtonActive(Button button){
        button.setBackgroundTintList(ColorStateList.valueOf(getColor(R.color.purple_700)));
    }

    public void setButtonInactive(Button button){
        button.setBackgroundTintList(ColorStateList.valueOf(getColor(R.color.grey)));
    }

    private void startEditor(){
        checkOverlayPermission();
        if(Settings.canDrawOverlays(this)) {
            startActivity(new Intent(this, EditorUI.class));
        }
    }

    private void startServer(boolean autorun){
        checkOverlayPermission();
        if(Settings.canDrawOverlays(this)) {
            server.setupServer();
            if (autorun) {
                new Thread(server::startServer).start();
            }
        }
    }

    private void checkOverlayPermission(){
        if (!Settings.canDrawOverlays(this)) {
            // send user to the device settings
            Intent myIntent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION);
            startActivity(myIntent);
        }
    }

}

File: app/src/main/java/com/xtr/keymapper/MainActivity.kt
package com.xtr.keymapper

import android.content.Intent
import android.content.res.ColorStateList
import android.os.Bundle
import android.provider.Settings
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {
    private var pointerOverlay: TouchPointer? = null
    @JvmField
    var server: Server? = null
    private var startOverlayButton: Button? = null
    private var startServerButton: Button? = null
    private var startInTerminal: Button? = null
    private var keymap: Button? = null
    private var configureButton: Button? = null
    private var infoButton: Button? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        server = Server(this)
        pointerOverlay = TouchPointer(this)
        Thread { pointerOverlay!!.startSocket() }.start()
        initButtons()
        setupButtons()
    }

    private fun setupButtons() {
        startServerButton!!.setOnClickListener { startServer(true) }
        startInTerminal!!.setOnClickListener { startServer(false) }
        startOverlayButton!!.setOnClickListener { startService() }
        keymap!!.setOnClickListener { startEditor() }
        configureButton!!.setOnClickListener {
            startActivity(
                Intent(
                    this,
                    InputDeviceSelector::class.java
                )
            )
        }
        infoButton!!.setOnClickListener {
            startActivity(
                Intent(
                    this,
                    InfoActivity::class.java
                )
            )
        }
    }

    private fun initButtons() {
        startOverlayButton = findViewById(R.id.startPointer)
        startServerButton = findViewById(R.id.startServer)
        startInTerminal = findViewById(R.id.startServerM)
        keymap = findViewById(R.id.start_editor)
        configureButton = findViewById(R.id.config_pointer)
        infoButton = findViewById(R.id.about_button)
    }

    private fun startService() {
        checkOverlayPermission()
        if (Settings.canDrawOverlays(this)) {
            pointerOverlay!!.open()
        }
    }

    fun setButtonActive(button: Button) {
        button.backgroundTintList = ColorStateList.valueOf(getColor(R.color.purple_700))
    }

    fun setButtonInactive(button: Button) {
        button.backgroundTintList = ColorStateList.valueOf(getColor(R.color.grey))
    }

    private fun startEditor() {
        checkOverlayPermission()
        if (Settings.canDrawOverlays(this)) {
            startActivity(Intent(this, EditorUI::class.java))
        }
    }

    private fun startServer(autorun: Boolean) {
        checkOverlayPermission()
        if (Settings.canDrawOverlays(this)) {
            server!!.setupServer()
            if (autorun) {
                Thread { server!!.startServer() }.start()
            }
        }
    }

    private fun checkOverlayPermission() {
        if (!Settings.canDrawOverlays(this)) {
            // send user to the device settings
            val myIntent = Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION)
            startActivity(myIntent)
        }
    }

    companion object {
        const val DEFAULT_PORT = 6234
        const val DEFAULT_PORT_2 = 6345
    }
}